### PR TITLE
UCP/AM: Added implementation of multi-fragment bcopy.

### PR DESCRIPTION
**Pull request by rakhmets**
## What
Added AM multi-fragment bcopy protocol for new infrastructure. 

## Why ?
The changes are part of UCP transition to new infrastructure.

### CODE REVIEW

File: src/ucp/am/eager_multi.c [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+static UCS_F_ALWAYS_INLINE ucs_status_t ucp_proto_eager_am_bcopy_multi_send_func(
+        ucp_request_t *req, const ucp_proto_multi_lane_priv_t *lpriv,
+        ucp_datatype_iter_t *next_iter)
+{
+    return ucp_proto_eager_bcopy_multi_common_send_func(
+            req, lpriv, next_iter, UCP_AM_ID_FIRST,
+            ucp_am_bcopy_pack_args_first, UCP_AM_FIRST_FRAG_META_LEN,
+            UCP_AM_ID_MIDDLE, ucp_am_bcopy_pack_args_mid,
+            UCP_AM_MID_FRAG_META_LEN);
```
Comment (brminich) [inline src/ucp/am/eager_multi.c:123]: Do we need to count `ucp_am_hdr_t` here as well?

File: test/gtest/ucp/test_ucp_am.cc [hunk 2]
```diff
@@ -1014,7 +1014,7 @@ UCS_TEST_P(test_ucp_am_nbx_eager_data_release, single_zcopy, "ZCOPY_THRESH=0")
     test_data_release(fragment_size() / 2);
 }
 
-UCS_TEST_SKIP_COND_P(test_ucp_am_nbx_eager_data_release, multi, enable_proto())
+UCS_TEST_P(test_ucp_am_nbx_eager_data_release, multi)
```
Comment (brminich) [inline test/gtest/ucp/test_ucp_am.cc:1017]: i think better to rename it to `multi_bcopy` and set "ZCOPY_THRESH=0", so you can later add bcopy

File: src/ucp/am/eager_multi.c (hunk 1 above)
Comment (rakhmets) [reply src/ucp/am/eager_multi.c:123]: It is taken into account:
```C
#define UCP_AM_FIRST_FRAG_META_LEN (sizeof(ucp_am_hdr_t) + sizeof(ucp_am_first_ftr_t))
#define UCP_AM_MID_FRAG_META_LEN (sizeof(ucp_am_hdr_t) + sizeof(ucp_am_mid_ftr_t))
```

Comment (brminich) [reply src/ucp/am/eager_multi.c:123]: right

**Review by brminich — APPROVED** [review]

File: src/ucp/Makefile.am [hunk 3]
```diff
@@ -161,5 +162,5 @@ libucp_la_SOURCES = \
 	wireup/wireup_cm.c \
 	stream/stream_send.c \
 	stream/stream_recv.c \
-	am/eager.c
-
+	am/eager.c \
```
Comment (yosefe) [inline src/ucp/Makefile.am:165]: should it be eager_single.c?

File: src/ucp/am/eager.c [hunk 4]
```diff
@@ -7,6 +7,7 @@
 #  include "config.h"
 #endif
 
+#include "eager.inl"
```
Comment (yosefe) [inline src/ucp/am/eager.c:10]: space line after

File: src/ucp/am/eager.inl [hunk 5]
```diff
@@ -0,0 +1,67 @@
+/**
+ * Copyright (C) 2022, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef AM_EAGER_INL_
+#define AM_EAGER_INL_
+
```
Comment (yosefe) [inline src/ucp/am/eager.inl:8]: remove one spaceline

File: src/ucp/am/eager_multi.c [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+ * Copyright (C) 2022, NVIDIA CORPORATION & AFFILIATES. ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "eager.inl"
+#include <ucp/proto/proto.h>
```
Comment (yosefe) [inline src/ucp/am/eager_multi.c:11]: one spaceline before

File: src/ucp/am/eager_multi.c [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+static size_t ucp_am_bcopy_pack_args_first(void *dest, void *arg)
+{
+    ucp_am_hdr_t *hdr                    = dest;
+    ucp_proto_multi_pack_ctx_t *pack_ctx = arg;
+    ucp_request_t *req                   = pack_ctx->req;
+    ucp_am_first_ftr_t *first_ftr;
+    size_t length, max_length;
+
+    ucs_assert(req->send.state.dt_iter.offset == 0);
```
Comment (yosefe) [inline src/ucp/am/eager_multi.c:80]: ucs_assertv to print offset

**Review by yosefe — APPROVED** [review]
pls squash and rebase

**Pull request merged**